### PR TITLE
DP-24803: Fix page ready event for leaflet map in Backstop

### PR DESCRIPTION
**Pull request by arthurbaghdas**
**Description:**
Explain the technical implementation of the work done.

**Jira:** (Skip unless you are MA staff)
DP-24803

**To Test:**
- [ ] Add steps to test this feature

---

[Peer Review Checklist](https://github.com/massgov/openmass/blob/develop/docs/peer_review_checklist.md)

### CODE REVIEW

File: backstop/scripts/ready.js [hunk 1]
```diff
@@ -113,6 +113,33 @@ module.exports = async function (page, scenario, vp) {
     await page.waitForSelector('.cbFormErrorMarker', {visible: true})
   }
 
+  let leafletMapInitialized = await page.evaluate(async function () {
+    let initialized = undefined;
+    var containers = document.querySelectorAll(".js-leaflet-map");
```
Comment (clairesunstudio) [inline backstop/scripts/ready.js]:
```suggestion
    const containers = document.querySelectorAll(".js-leaflet-map");
```

File: backstop/scripts/ready.js [hunk 2]
```diff
@@ -113,6 +113,33 @@ module.exports = async function (page, scenario, vp) {
...
+    if (containers.length) {
+      containers.forEach(function (e) {
+        var container = L.DomUtil.get(e);
```
Comment (clairesunstudio) [inline backstop/scripts/ready.js]:
```suggestion
        const container = L.DomUtil.get(e);
```

File: backstop/scripts/ready.js [hunk 3]
```diff
@@ -113,6 +113,33 @@ module.exports = async function (page, scenario, vp) {
...
+        if (container != null) {
```
Comment (clairesunstudio) [inline backstop/scripts/ready.js:122]: can you add some notes to these conditions?  

I feel like the condition should be reserved - currently, `if (containers.length)` is false, then `initialized` is set to true - that means for pages that don't have leaflet maps, we are adding a 5 second timeout ?

File: backstop/scripts/ready.js [hunk 4]
```diff
@@ -113,6 +113,33 @@ module.exports = async function (page, scenario, vp) {
...
+        }
+      });
+    }
+    initialized = true;
+  })
+
+  if (leafletMapInitialized == false) {
+    await page.waitForSelector('.js-leaflet-map .leaflet-tile-container', {
+      visible: true,
+      timeout: 10000,
```
Comment (clairesunstudio) [inline backstop/scripts/ready.js:137]:
```suggestion
      timeout: 10000,
```
we are adding a 10-second timeout here? is that too much?

File: backstop/scripts/ready.js [hunk 5]
```diff
@@ -113,6 +113,33 @@ module.exports = async function (page, scenario, vp) {
...
 
+  let leafletMapInitialized = await page.evaluate(async function () {
+    let initialized = undefined;
+    var containers = document.querySelectorAll(".js-leaflet-map");
+    if (containers.length) {
+      containers.forEach(function (e) {
+        var container = L.DomUtil.get(e);
+        if (container != null) {
+          if (container._leaflet_id == null) {
+            initialized = false;
```
Comment (clairesunstudio) [inline backstop/scripts/ready.js:124]:
```suggestion
            initialized = false;
```

I'm a little confused by the logic here - by chaining these conditions `containers.length` && `container != null` && `container._leaflet_id == null` means we are only setting `initialized` to false when all three conditions are met, and if any is unmet we are setting `initialized` to true?

File: backstop/scripts/ready.js (hunk 4 above)
Comment (arthurbaghdas) [reply backstop/scripts/ready.js:137]: That's the maximum time that we wait for an element to become visible, it is usually less then 2 second.

File: backstop/scripts/ready.js (hunk 5 above)
Comment (arthurbaghdas) [reply backstop/scripts/ready.js:124]: If the map is not initialized   container._leaflet_id returns null, that's the condition we need to establish a wait for the map to become visible

File: backstop/scripts/ready.js (hunk 3 above)
Comment (arthurbaghdas) [reply backstop/scripts/ready.js:122]: We are not running this on the pages that don't have a map.

**Comment by clairesunstudio** [comment]
redeployed to feature1 and ran backstop --> https://output.circle-artifacts.com/output/job/373b0e50-ee80-4d04-af64-08bfe8b78b64/artifacts/0/home/circleci/code/backstop/report/index.html
none of the diffs are related to maps. Approved!

**Pull request merged**